flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultVertexParallelismDecider.java

```diff
@@ -109,54 +116,90 @@ private int calculateParallelism(List<BlockingResultInfo> consumedResults) {
 
             broadcastBytes = expectedMaxBroadcastBytes;
         }
 
-        int parallelism =
+        int initialParallelism =
                 (int) Math.ceil((double) nonBroadcastBytes / (dataVolumePerTask - broadcastBytes));
+        int parallelism = normalizeParallelism(initialParallelism);
 
         LOG.debug(
                 "The size of broadcast data is {}, the size of non-broadcast data is {}, "
-                        + "the initially decided parallelism is {}.",
+                        + "the initially decided parallelism is {}, after normalize is {}",
                 new MemorySize(broadcastBytes),
                 new MemorySize(nonBroadcastBytes),
+                initialParallelism,
                 parallelism);
 
         if (parallelism < minParallelism) {
             LOG.info(
-                    "The initially decided parallelism {} is smaller than the minimum parallelism {} "
-                            + "(which is configured by '{}'). Use {} as the finally decided parallelism.",
+                    "The initially normalized parallelism {} is smaller than the normalized minimum parallelism {}. "
+                            + "Use {} as the finally decided parallelism.",
                     parallelism,
                     minParallelism,
-                    JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_MIN_PARALLELISM.key(),
                     minParallelism);
             parallelism = minParallelism;
         } else if (parallelism > maxParallelism) {
             LOG.info(
-                    "The initially decided parallelism {} is larger than the maximum parallelism {} "
-                            + "(which is configured by '{}'). Use {} as the finally decided parallelism.",
+                    "The initially normalized parallelism {} is larger than the normalized maximum parallelism {}. "
+                            + "Use {} as the finally decided parallelism.",
                     parallelism,
                     maxParallelism,
-                    JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_MAX_PARALLELISM.key(),
                     maxParallelism);
             parallelism = maxParallelism;
         }
 
         return parallelism;
     }
 
-    public static DefaultVertexParallelismDecider from(Configuration configuration) {
+    @VisibleForTesting
+    int getMaxParallelism() {
+        return maxParallelism;
+    }
+
+    @VisibleForTesting
+    int getMinParallelism() {
+        return minParallelism;
+    }
+
+    static DefaultVertexParallelismDecider from(Configuration configuration) {
+        int maxParallelism = getNormalizedMaxParallelism(configuration);
+        int minParallelism = getNormalizedMinParallelism(configuration);
+        checkState(
+                maxParallelism >= minParallelism,
+                String.format(
+                        "You should adjust '%s' and '%s' so that there is at least one power of 2 between them.",
```

Comment:
>> You should adjust '%s' and '%s' so that there is at least one power of 2 between them.

Error can also happen if max < min.
Maybe "{max-parallelism} should be greater than or equal to {min-parallelism} and the range must contain at least one power of 2."?